### PR TITLE
Add fuelTime to Fluids

### PR DESCRIPTION
**Pull request by LemmaEOF**
At the moment, if you want to make something fueled by combustible fluids (like Thermal Expansion's Combustion Dynamo or Immersive Engineering's Diesel Generator), you need to either provide your own fluid fuels or explicitly code in each external fluid, how fast it's consumed, and how much power you get out of it.

This pull request adds a new property to Fluids, called `fuelTime`. This functions exactly like the `itemBurnTime` for furnace fuel, being the amount of ticks taken to consume 1 bucket of fuel. A boolean property is added called `isCombustible()` which returns true if `fuelTime` is greater than 0, to be used for input validators. Anything with a `fuelTime` of 0 is considered non-combustible and should not be accepted into fluid-fueled generators.

The advantage of the Fluid fuelTime system is that it removes the need to adapt for any mod that adds in edge cases of combustible fuels; a dev can just use `getFuelTime()` like they would `TileEntityFurnace.getItemBurnTime(usedFuel());`. Almost all mods that have fluid-fueled machines have a furnace-like solid-fuel machine, so code from the solid-fuel machine can easily be adapted to fluid-fuel. For example, this is the code I use for fuel consumption in a solid-fueled generator I have:
```java
private boolean consumeFuel() {
        if (currentFuelTime <= 0) {
            ItemStack usedFuel = inv.extractItem(SLOT_FUEL, 1, false);
            if (!usedFuel.isEmpty()) {
                int newFuelTicks = TileEntityFurnace.getItemBurnTime(usedFuel);
                maxFuelTime = newFuelTicks;
                currentFuelTime = newFuelTicks;
            } else {
                return false;
            }
        }
        currentFuelTime--;
        return true;
    }
```

And this is the code adapted for fluid fuel:
```java
private boolean consumeFuel() {
        if (currentFuelTime <= 0) {
            FluidStack usedFuel = tank.drain(1000, true);
            if (usedFuel != null) {
                int newFuelTicks = tank.getFluid().getFuelTime();
                maxFuelTime = newFuelTicks;
                currentFuelTime = newFuelTicks;
            } else {
                return false;
            }
        }
        currentFuelTime--;
        return true;
    }
```

Adding the fuelTime property *is not a breaking change*, so there are no downsides to adding it to Forge. Mods are free to use it if they want, and since it's such a simple change, it's easy for devs to implement safely.

### CODE REVIEW

**Comment by covers1624** [comment]
Looks good. I would recommend an overload to `getFuelTime` that takes a `FluidStack` parameter like all the other properties in `Fluid`. Also for `isCombustible` the same should be done, overload with `FluidStack` argument.

**Comment by mezz** [comment]
As discussed on IRC, this looks good to me.
`FluidStack` version is essential because fluids are weird.

**Comment by jabelar** [comment]
This looks good. Actually, in the implementation example for more "realism" I would personally consume the fuel as each tick went along (i.e. reduce the fluid quantity by 1000/fuelTime each tick) rather than draining a bucket only after fuel time expires.

I have been similarly thinking to make a PR for a combustibility property of whether the fluid can burn in the world or alternatively extinguish fire in the world. I think this could be extended to that. In other words, after this PR is implemented I will probably extend the use of isCombustible() to check whether a fluid can extinguish fire or alternatively catch on fire.

I like all the PRs related to fluids. Seems like a very popular category for modding, and I've got a major PR pending (hint hint to the PR approval gods) that will give modders a lot of control over which water-like properties a custom fluid has.  I think my mod plus this PR and the PR for fluid ingredients and we'll be very far along with a comprehensive fluid system.

File: src/main/java/net/minecraftforge/fluids/Fluid.java [hunk 1]
```diff
@@ -104,6 +104,18 @@
...
      */
     protected int viscosity = 1000;
+	
+    /**
+     * Amount of time a fluid will burn for in a fluid-fueled generator.
+     * Measured in fuel ticks per bucket of fluid.
+     * 
+     * Default value is 0; anything above is considered a combustible fluid.
+     * Hot fluids are hot, not combustible, so their value should be 0.
+     * If fuelEnergy > 0, isCombustible will return true.
```
Comment (bs2609) [inline src/main/java/net/minecraftforge/fluids/Fluid.java]: Missed a rename here

**Comment by KnightMiner** [comment]
You should override getItemBurnTime in the Forge universal bucket to check the fuel temp and use that to determine a buckets fuel time so this has a use by default. You should probably also give lava a burn time so modders do not need to special case it.

**Comment by jabelar** [comment]
The suggestion about using fuel temperature for this brings up a point. The word "combustible" means it can actually be burned (catch on fire and consumed). However, for cooking/smelting purposes a fuel that burns is good but also a fuel that is just very hot like lava. Lava doesn't really burn or combust, it simply is super hot.

So to be truly semantically correct, temperature, combustibility and the use as "fuel" are all not interchangeable. For example, cold gasoline can burn and hot water can will never burn (but it can cook).

For that reason I would personally prefer the name of the method to be isFuel() or similar. I don't feel super strongly about it, but like accuracy in naming and felt compelled to point it out.

**Comment by LemmaEOF** [comment]
@KnightMiner This PR is specifically for combustible fluids, fluids that will burn if you set them on fire. It's so modders don't have to make edge cases for every version of oil, petrol, kerosene, gasoline, biogas, diesel, etc. from every mod be compatible with their fluid-fuel generators or machines. As I say in the javadocs for fuelTime, hot fluids are hot, not combustible.

Being able to use lava in these sorts of machines would break a *lot* of balancing, since it is very easy to get practically infinite lava relatively early in the game. Save for BuildCraft oil, any combustible fuel needs to be processed at least once to be usable, and combustion-based machines are balanced around that.

**Comment by KingLemming** [comment]
So my reservations with this are simply that fluid fuels aren't nearly as standardized as solid ones. And that is *not a bad thing.*

Your statements that it can be added without breaking things are true, except once it is "known" about, a lot of issues are going to be opened up across a lot of repositories complaining about balance one way or the other, or why YOUR MOD DEVIATES FROM THE REAL ACTUAL VALUE OF THE FLUID RAR WHAT IS WRONG WITH YOU STUPID MODDER!!!????11!!?

I say this from experience. And I think there will actually be *extremely* few cases where this burn time is useful, because burn time will only correlate to RF (or FE, whatever) in a 10:1 ratio. So then there's ambiguity about units, as basically...nothing in all of modded MC will make use of a furnace-style "burnTime" value.

**Comment by LemmaEOF** [comment]
@KingLemming, would it be better if it was a value for how much RF/FE/etc. should be gotten out of a bucket of that fuel instead? I made it a fuelTime so mods could decide how efficient their generators were, but you do make the very good point that leaving how much rf you get per bucket to the hands of individual modders isn't always the best idea.

Since it's just a name right now, it would be very easy for me to just change it from `fuelTime` to `fuelEnergy` and update the javadocs to describe it properly.

**Comment by alexbobp** [comment]
I think almost every tech mod has some kind of solid fuel based generator and thus some concept of what traditional furnace fuel ticks are worth.

**Comment by jabelar** [comment]
For balancing issues, just make sure that the fuel time property has a setter method. Mods that that work on top of other mods but don't like the balance could then adjust everything.

At some point we shouldn't over-think this. The main point is to have control of  the conversion of fluid quantity to "energy". That is useful. The problem with balancing already exists and isn't made any worse by this.

**Comment by KingLemming** [comment]
So there's actually another quirk that I hadn't initially thought of - canonical fluids. Fluids at the moment don't have delegates. They aren't handled like other things, so a world will get a "canonical" version. This means that you'll get some possibly inconsistent behavior with the same modlist depending on when the mods were installed.

Annoying, I know, and *typically* it doesn't matter, but when you get into fuel values and different mods have different ideas on what "the" value should be...ehh...

**Comment by LemmaEOF** [comment]
...wow, fluids really need to be fixed at a fundamental level. If I had the time I'd write up something to fix all the issues fluids have, but this is all I've got at the moment.

**Comment by KingLemming** [comment]
Well I wrote them back in 1.6 and they've barely been touched. So that's what...4 1/2 years ago? I think the only updates of note were capabilities, and even that was more or less passthrough rewrites of existing stuff.

**Comment by HenryLoenwind** [comment]
As someone who's responsible for a mod that burns fluids: In contrast to solid fuels, that are always implemented to mimic vanilla's furnace with just a burnTime value, fluids often go with time and "hotness". So you'll often see `"burn time in ticks"*"RF/t"="RF total"` calculations. Some mods determine the RF/t by machine, and get burnTime and RFtotal from the fluid, some determine burnTime by machine and get RF/t and RFtotal from the fluid. However, that difference doesn't matter as those two cases can be calculated from each other as long as the fluid provides any two values.

tl;DR: Just burnTime isn't enough, give either time+energy/t or time+totalEnergy.

Otherwise good idea. Plenty of ways to fail. But at least it's a some minimal common ground.

**Comment by jabelar** [comment]
I think providing a public setter solves most of these issues. If your mod has a formula for such things there is nothing stopping it from adjusting the default upon use, or if your mod has different view on the balance.

Forge can't solve balance issues directly. There is never going to be anything that stops me from making a mod where my gasoline fuel is 10 times easier to get then your gasoline, or where my gasoline provides more energy than your uranium. That sort of stuff should be sorted out by discussions between mod authors of mods that people like to combine in mod packs.

The discussion here should just be to make a common interface for setting and adjusting. That will go a long way. I think the main thing to discuss is whether we make the interface about the time, or about the total "energy".

But please let's not make this a "let's unify all gasoline" discussion. There is already PR #4457 to address unification in general. This PR should provide the hooks where modders can work together to come up with community-accepted values.

**Comment by liach** [comment]
@Boundarybreaker I strongly suggest adding a mechanism for other mods to override these fuel values

**Comment by KingLemming** [comment]
@liach - frankly, I will likely be ignoring them, and I suspect most modders will too. That's kind of the problem, but it's the only way we can ensure our own mods are balanced and don't get ninja-wrecked by a mod "doing it wrong." Most of us already have a ballpark figure on what a fluid should be worth, and we might be +/- 10%. Thing is, that's probably good enough.

Again, intention here is good, but fluids have deeper issues and I suspect we'd be better off holding this until 1.13 when they are more properly unified.

**Comment by LemmaEOF** [comment]
Ok, with Mezz creating #4693 as a fix for fluids doFill/doDrain for 1.13, would it be a good idea to just have 1.13 contain a full fluid overhaul and push this back to 1.13 as well?

**Comment by KingLemming** [comment]
I think so, yes.

**Comment by liach** [comment]
imo the fuel time shoulf have an accessor in fluidstack and should be nbt aware.

**Comment by gigaherz** [comment]
The fluid system redesign has just landed on the 1.14.x branch, and is ripe for getting this feature included in it, through the `FluidAttributes` builder.
Anyone up for porting this over to the new code?

**Comment by LemmaEOF** [comment]
I don't think I'm gonna port this, I'm burned out on modding.

**Comment by gigaherz** [comment]
No problem I'll integrate this into #6110

**Comment by gigaherz** [comment]
The changes are now in the PR, to be merged soon. I have chosen to take some liberties with the implementation: instead of only giving combustible fluids a time, I have made it a number that would include hot fluids, even lava. To exclude non-combustible fluids from combustion generators, a tag should be used, with my recommendation to use `forge:fluids/combustible`.
